xds/src/main/java/io/grpc/xds/AbstractXdsClient.java

```diff
@@ -0,0 +1,760 @@
+        }
+      });
+    }
+  }
+
+  protected enum ResourceType {
+    UNKNOWN, LDS, RDS, CDS, EDS;
+
+    String typeUrl() {
+      switch (this) {
+        case LDS:
+          return ADS_TYPE_URL_LDS;
+        case RDS:
+          return ADS_TYPE_URL_RDS;
+        case CDS:
+          return ADS_TYPE_URL_CDS;
+        case EDS:
+          return ADS_TYPE_URL_EDS;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown or missing case in enum switch: " + this);
+      }
+    }
+
+    String typeUrlV2() {
+      switch (this) {
+        case LDS:
+          return ADS_TYPE_URL_LDS_V2;
+        case RDS:
+          return ADS_TYPE_URL_RDS_V2;
+        case CDS:
+          return ADS_TYPE_URL_CDS_V2;
+        case EDS:
+          return ADS_TYPE_URL_EDS_V2;
+        case UNKNOWN:
+        default:
+          throw new AssertionError("Unknown or missing case in enum switch: " + this);
+      }
+    }
+
+    private static ResourceType fromTypeUrl(String typeUrl) {
+      switch (typeUrl) {
+        case ADS_TYPE_URL_LDS:
+          // fall trough
+        case ADS_TYPE_URL_LDS_V2:
+          return LDS;
+        case ADS_TYPE_URL_RDS:
+          // fall through
+        case ADS_TYPE_URL_RDS_V2:
+          return RDS;
+        case ADS_TYPE_URL_CDS:
+          // fall through
+        case ADS_TYPE_URL_CDS_V2:
+          return CDS;
+        case ADS_TYPE_URL_EDS:
+          // fall through
+        case ADS_TYPE_URL_EDS_V2:
+          return EDS;
+        default:
+          return UNKNOWN;
+      }
+    }
+  }
+
+  private abstract class AbstractAdsStream {
+    private boolean responseReceived;
+    private boolean closed;
+
+    // Response nonce for the most recently received discovery responses of each resource type.
+    // Client initiated requests start response nonce with empty string.
+    // A nonce is used to indicate the specific DiscoveryResponse each DiscoveryRequest
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+    private String cdsRespNonce = "";
+    private String edsRespNonce = "";
+
+    abstract void start();
+
+    abstract void sendError(Exception error);
+
+    /**
+     * Sends a discovery request with the given {@code versionInfo}, {@code nonce} and
+     * {@code errorDetail}. Used for reacting to a specific discovery response. For
+     * client-initiated discovery requests, use {@link
+     * #sendDiscoveryRequest(ResourceType, Collection)}.
+     */
+    abstract void sendDiscoveryRequest(ResourceType type, String versionInfo,
+        Collection<String> resources, String nonce, @Nullable String errorDetail);
+
+    /**
+     * Sends a client-initiated discovery request.
+     */
+    final void sendDiscoveryRequest(ResourceType type, Collection<String> resources) {
+      String nonce;
```

Comment:
Add an INFO level log for sending discovery `type` request for `resources`? Just as the INFO levels log in NACK and ACK.

Reply:
Sure, added a log message.